pkg/stash/with_locker.go

```diff
@@ -0,0 +1,133 @@
+package stash
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/gomods/athens/pkg/errors"
+	"github.com/gomods/athens/pkg/observ"
+	"github.com/gomods/athens/pkg/storage"
+)
+
+const (
+	defaultGetLockTimeout = 5 * time.Minute
+	defaultPingInterval   = 10 * time.Second
+)
+
+type locker interface {
+	// lock obtains a lock to be used for single flight.
+	// The lock will be released when ctx is canceled. When the lock is released whether intentionally or not,
+	// releaseErrs will receive an error. If that error is nil, the lock was successfully released as a response to ctx.
+	// Returns:
+	//	 releaseErrs:	A channel that emits an error when the lock is released.
+	//   err: 			Any error creating the lock. If err is non-nil, no lock is created.
+	lock(ctx context.Context, name string) (releaseErrs <-chan error, err error)
```

Comment:
see my comment below. I think it would be a much easier API to work with if lock's signature was this: 

`lock(ctx context.Context, name string) (ctx context.Context, err error)` 

If we care about what the releaseErr is, we can mimic a mutext behavior by having both a `lock` and a `unlock` method and do the `defer mu.Unlock()`  similar to how the etcd implementation has been doing it. 

Let me know if im missing something :v: